internal/hud/server/websocket_reader.go

```diff
@@ -0,0 +1,174 @@
+// TODO(maia): figure out if this should live here, or elsewhere (I assume
+//   it's worth it to make a generic state-reader b/c we may want other cmds
+//   that interact with a running Tilt e.g. `tilt status`.)
+
+import (
+	"bytes"
+	"log"
+	"os"
+	"os/signal"
+	"time"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/gorilla/websocket"
+	"github.com/mattn/go-colorable"
+	"github.com/tilt-dev/tilt/internal/hud"
+	"github.com/tilt-dev/tilt/internal/hud/webview"
+	"github.com/tilt-dev/tilt/pkg/model"
+	"github.com/tilt-dev/tilt/pkg/model/logstore"
+	proto_webview "github.com/tilt-dev/tilt/pkg/webview"
+)
+
+// TODO: interface
+type WebsocketReader struct {
+	url     url.URL
+	handler ViewHandler
+}
+
+func ProvideWebsockerReader() *WebsocketReader {
+	return &WebsocketReader{
+		// TODO(maia): pass this URL instead of hardcoding / wire this
+		url:     url.URL{Scheme: "ws", Host: "localhost:10350", Path: "/ws/view"},
+		handler: NewLogStreamer(),
+	}
+}
+
+type ViewHandler interface {
+	Handle(v proto_webview.View) error
+}
+
+type LogStreamer struct {
+	logstore   *logstore.LogStore
+	printer    *hud.IncrementalPrinter
+	checkpoint logstore.Checkpoint
+}
+
+func NewLogStreamer() *LogStreamer {
+	// TODO(maia): wire this
+	printer := hud.NewIncrementalPrinter(hud.Stdout(colorable.NewColorableStdout()))
+	return &LogStreamer{
+		logstore: logstore.NewLogStore(),
+		printer:  printer,
+	}
+}
+func (ls *LogStreamer) Handle(v proto_webview.View) error {
+	fromCheckpoint := logstore.Checkpoint(v.LogList.FromCheckpoint)
+	toCheckpoint := logstore.Checkpoint(v.LogList.ToCheckpoint)
+
+	if fromCheckpoint == -1 {
+		// Server has no new logs to send
+		return nil
+	}
+
+	segments := v.LogList.Segments
+	if fromCheckpoint < ls.checkpoint {
+		// The server is re-sending some logs we already have, so slice them off.
+		deleteCount := ls.checkpoint - fromCheckpoint
+		segments = segments[deleteCount:]
+	}
+
+	// TODO(maia): filter for the resources that we care about (`tilt logs resourceA resourceC`)
+	//   --> and if there's only one resource, don't prefix logs with resource name?
+	for _, seg := range segments {
+		// TODO(maia): secrets???
+		ls.logstore.Append(webview.LogSegmentToEvent(seg, v.LogList.Spans), model.SecretSet{})
+	}
+
+	ls.printer.Print(ls.logstore.ContinuingLines(ls.checkpoint))
+
+	if toCheckpoint > ls.checkpoint {
+		ls.checkpoint = toCheckpoint
+	}
+
+	return nil
+}
+
+func (wsr *WebsocketReader) Listen() {
+	// catch signals to kill
+	// TODO: use signal catching we already use in `up` etc.
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt)
+
+	log.Printf("connecting to %s", wsr.url.String())
+
+	c, _, err := websocket.DefaultDialer.Dial(wsr.url.String(), nil)
+	if err != nil {
+		log.Fatal("dial:", err)
+	}
+	defer c.Close()
+
+	done := make(chan struct{})
+
+	go func() {
+		// TODO(maia): make sure this closes okay 😅
+		defer close(done)
+		for {
+			_, data, err := c.ReadMessage()
```

Comment:
idiomatically you'd usually do c.NextReader(), and unmarshal directly off the reader (for some reasons as using NextWriter below)